cmd/promtool/sd.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/relabel"
+)
+
+type sdCheckResult struct {
+	targetGroups   []*targetgroup.Group
+	relabelConfigs []*relabel.Config
+}
+
+// CheckSD performs service discovery for the given job name and reports the results.
+func CheckSD(sdConfigFiles, sdJobName string, sdTimeout time.Duration) int {
+	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+
+	cfg, err := config.LoadFile(sdConfigFiles, false, logger)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Cannot load config", err)
+		return 2
+	}
+
+	var discoveryConfigs discovery.Configs
+	var relabelConfigs []*relabel.Config
+	jobs := []string{}
+	jobMatched := false
+	for _, v := range cfg.ScrapeConfigs {
+		jobs = append(jobs, v.JobName)
+		if v.JobName == sdJobName {
+			jobMatched = true
+			discoveryConfigs = v.ServiceDiscoveryConfigs
+			relabelConfigs = v.RelabelConfigs
+			break
+		}
+	}
+
+	if !jobMatched {
+		fmt.Fprintf(os.Stderr, "Job %s not found. Select one of:\n", sdJobName)
+		for _, job := range jobs {
+			fmt.Fprintf(os.Stderr, "\t%s\n", job)
+		}
+		return 1
+	}
+
+	targetGroupChan := make(chan []*targetgroup.Group)
+	ctx, cancel := context.WithTimeout(context.Background(), sdTimeout)
+	defer func() {
+		cancel()
+		close(targetGroupChan)
+	}()
+
+	for _, cfg := range discoveryConfigs {
+		d, err := cfg.NewDiscoverer(discovery.DiscovererOptions{Logger: logger})
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "Could not create new discoverer", err)
+			return 2
+		}
+		go d.Run(ctx, targetGroupChan)
+	}
+
+	var targetGroups []*targetgroup.Group
+	returnCode := 0
+	select {
+	case <-ctx.Done():
+		fmt.Fprintln(os.Stderr, "Timeout exceeded")
+		returnCode = 2
+	case targetGroups = <-targetGroupChan:
+		fmt.Println(sdCheckResult{targetGroups: targetGroups, relabelConfigs: relabelConfigs})
+		break
+	}
+
+	return returnCode
+}
+
+func (res sdCheckResult) String() string {
+	result := ""
```

Comment:
Use strings.Builder, means you can do `fmt.Fprintf(&result, ...)` rather than `result += fmt.Sprintf(...)`` and will look cleaner.

Reply:
I would prefer that we pretty print all this as json rather than complex formatting like this.

Reply:
using json.marshal now...